File: backend/app/Models/AppNotification.php
<?php

namespace App\Models;

use App\Models\Concerns\BelongsToTenant;
use Database\Factories\AppNotificationFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * tenant_id is fillable here (unlike most tenant-owned models) for the
 * same reason as TenantSubscription/PaymentTransaction: rows are
 * created from services (NotificationService, AppNotificationChannel),
 * including from console/scheduled commands with no ambient tenant
 * context to auto-fill from — see BelongsToTenant's creating hook. No
 * FormRequest ever exposes this field.
 */
#[Fillable(['tenant_id', 'type', 'title', 'message', 'read', 'user_id', 'member_id'])]
class AppNotification extends Model
{
    /** @use HasFactory<AppNotificationFactory> */
    use BelongsToTenant, HasFactory;

    protected $table = 'notifications';

    protected function casts(): array
    {
        return [
            'read' => 'boolean',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /** A member-facing notification (Phase 25) — mutually exclusive with user_id, see the migration. */
    public function member(): BelongsTo
    {
        return $this->belongsTo(Member::class);
    }
}

File: mobile/lib/shared/models/app_notification.dart
class AppNotification {
  const AppNotification({
    required this.id,
    required this.type,
    required this.title,
    required this.message,
    required this.read,
    required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) => AppNotification(
        id: json['id'] as int,
        type: json['type'] as String? ?? 'System Notification',
        title: json['title'] as String? ?? '',
        message: json['message'] as String? ?? '',
        read: json['read'] as bool? ?? false,
        createdAt: json['createdAt'] as String? ?? '',
      );

  final int id;
  final String type;
  final String title;
  final String message;
  final bool read;
  final String createdAt;

  AppNotification copyWith({bool? read}) => AppNotification(
        id: id,
        type: type,
        title: title,
        message: message,
        read: read ?? this.read,
        createdAt: createdAt,
      );
}
